xdsl/dialects/riscv_func.py

```diff
@@ -48,9 +51,102 @@ def verify_(self):
             )
 
 
+@irdl_op_definition
+class CallOp(IRDLOperation):
+    """RISC-V function call operation"""
+
+    name = "riscv_func.call"
+    args: Annotated[VarOperand, riscv.RegisterType]
+    func_name: OpAttr[StringAttr]
+    result: Annotated[OptOpResult, riscv.RegisterType]
+
+    def __init__(
+        self,
+        func_name: StringAttr,
+        args: list[Operation | SSAValue],
+        has_result: bool = True,
+        comment: StringAttr | None = None,
+    ):
+        super().__init__(
+            operands=[args],
+            result_types=[riscv.RegisterType(riscv.Register()) if has_result else None],
+            attributes={
+                "func_name": func_name,
+                "comment": comment,
+            },
+        )
+
+
+@irdl_op_definition
+class FuncOp(IRDLOperation):
+    """RISC-V function definition operation"""
+
+    name = "riscv_func.func"
+    args: Annotated[VarOperand, riscv.RegisterType]
+    func_name: OpAttr[StringAttr]
+    func_body: SingleBlockRegion
+
+    def __init__(self, name: str, region: Region):
+        attributes: dict[str, Attribute] = {"func_name": StringAttr(name)}
+
+        super().__init__(attributes=attributes, regions=[region])
+
+    def verify_(self):
+        # Check that the returned value matches the type of the function
+        if len(self.func_body.blocks) != 1:
+            raise VerifyException("Expected FuncOp to contain one block")
+
+        block = self.func_body.blocks[0]
+
+        if not len(block.ops):
+            raise VerifyException("Expected FuncOp to not be empty")
+
+        last_op = block.last_op
+
+        if not isinstance(last_op, ReturnOp):
+            raise VerifyException("Expected last op of FuncOp to be a ReturnOp")
+
+        if len(self.args) >= 9:
```

Comment:
Should we put a `#TODO` here for when register allocation implements register spilling? @superlopuh ?

Reply:
I feel that register allocation should be done *before* this is lowered, right? Otherwise it's gonna be a lot more difficult to tell where function calls are/should be and where to increment/decrement stack pointers...

Reply:
My understanding is that we would want to do it before and after, probably. There's some information that is easy to have at the function level that becomes a bit harder to kep track of when we lower to jumps and rets. But there might be a bit more to do after. Our mixing of allocated and unallocated registers would make it easier to try a few approaches.